java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -102,3 +104,97 @@ class MyBatisSqlOperationAnnotationMethod extends Method {
+  string taintedArgs;
+  string signature;
+
+  MyBatisAbstractSqlMethod() {
+    this.getDeclaringType().getSourceDeclaration() instanceof MyBatisAbstractSql and
+    (
+      this.hasName([
+          "UPDATE", "SET", "INSERT_INTO", "SELECT", "OFFSET_ROWS", "LIMIT", "OFFSET",
+          "FETCH_FIRST_ROWS_ONLY", "DELETE_FROM", "INNER_JOIN", "ORDER_BY", "WHERE", "HAVING",
+          "OUTER_JOIN", "LEFT_OUTER_JOIN", "RIGHT_OUTER_JOIN", "GROUP_BY", "FROM", "SELECT_DISTINCT"
+        ]) and
+      taintedArgs = "Argument[0]" and
+      signature = "String"
+      or
+      this.hasName([
+          "SET", "INTO_COLUMNS", "INTO_VALUES", "SELECT_DISTINCT", "FROM", "JOIN", "INNER_JOIN",
+          "LEFT_OUTER_JOIN", "RIGHT_OUTER_JOIN", "OUTER_JOIN", "WHERE", "GROUP_BY", "HAVING",
+          "ORDER_BY"
+        ]) and
+      taintedArgs = "Argument[0].ArrayElement" and
+      signature = "String[]"
+      or
+      this.hasName("VALUES") and taintedArgs = "Argument[0..1]" and signature = "String,String"
+    )
+  }
+
+  string getTaintedArgs() { result = taintedArgs }
+
+  string getCsvSignature() { result = signature }
+}
+
+/**
+ * A return statement of a method used in a MyBatis Provider.
+ *
+ * See
+ * - `MyBatisProvider`
+ * - https://mybatis.org/mybatis-3/apidocs/org/apache/ibatis/annotations/package-summary.html
+ */
+class MyBatisInjectionSink extends DataFlow::Node {
+  MyBatisInjectionSink() {
+    exists(Annotation a, Method m |
+      a.getType() instanceof MyBatisProvider and
+      m.getDeclaringType() = a.getValue(["type", "value"]).(TypeLiteral).getTypeName().getType() and
+      m.hasName(a.getValue("method").(StringLiteral).getValue()) and
+      this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+    )
+  }
+}
+
+private class MyBatisProviderStep extends TaintTracking::AdditionalValueStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(MethodAccess ma, Annotation a, Method providerMethod |
+      exists(int i |
+        ma.getArgument(i) = n1.asExpr() and
+        providerMethod.getParameter(i) = n2.asParameter()
+      )
+    |
+      a.getType() instanceof MyBatisProvider and
+      ma.getMethod().getAnAnnotation() = a and
+      providerMethod.getDeclaringType() =
+        a.getValue(["type", "value"]).(TypeLiteral).getTypeName().getType() and
+      providerMethod.hasName(a.getValue("method").(StringLiteral).getValue())
+    )
+  }
+}
+
+private class MyBatisAbstractSqlToStringStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row = "org.apache.ibatis.jdbc;AbstractSQL;true;toString;;;Argument[-1];ReturnValue;taint"
+  }
+}
+
+private class MyBatisAbstractSqlMethodsStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    exists(MyBatisAbstractSqlMethod m |
+      row =
+        "org.apache.ibatis.jdbc;AbstractSQL;true;" + m.getName() + ";(" + m.getCsvSignature() +
+          ");;" + m.getTaintedArgs() + ";Argument[-1];taint"
+    )
```

Comment:
This isn't really the way we want to add csv-rows to our models-as-data coverage. As-is this is an awkward mix of traditional QL-modelling and a csv string bits.
We would like our models-as-data to be database independent, such that these rows don't depend on the existence of the `MyBatisAbstractSql` type.

Reply:
Done in https://github.com/github/codeql/pull/8345/commits/9e1b98e5a4da22c60e36871dc2486d16f91555df. Would you prefer having the rows like https://github.com/github/codeql/commit/f9ae1efa2c3848a4ddf711c4902d28ad598d87c7#diff-34610297bc46b266af677a9ce54f188299bf011b8b4ec4b04264fab2231ded21R162?

Reply:
I would prefer the latter.